code/xcvm/lib/core/src/proto.rs

```diff
@@ -360,6 +382,34 @@ where
 	}
 }
 
+impl<TAbiEncoded, TAccount, TAssets> TryFrom<Exchange>
+	for crate::Instruction<TAbiEncoded, TAccount, TAssets>
+where
+	TAbiEncoded: TryFrom<Vec<u8>>,
+	TAccount: for<'a> TryFrom<&'a [u8]>,
+	TAssets: From<Vec<(crate::AssetId, crate::Balance)>>,
+{
+	type Error = ();
+
+	fn try_from(value: Exchange) -> core::result::Result<Self, Self::Error> {
+		Ok(crate::Instruction::Exchange {
+			id: value.id.and_then(|x| x.id).map(TryInto::try_into).ok_or(())?.map_err(|_| ())?,
```

Comment:
```suggestion
			id: value.id.and_then(|x| x.id).map(TryInto::try_into).ok_or(())?.map_err(|_| ())?,
			id: value.id.and_then(|msg| msg.id.try_into().ok()).ok_or(|_| ())?,
```

should also work I think and is a tad shorter.

Reply:
hm, did not compiled (